### PR TITLE
Nb feat/implement level db

### PR DESCRIPTION
**Pull request by nicobarra**
In the following PR I implemented an api that makes a CRUD on leveldb over a Person struct: 

* It has a router that creates the routes and executes the handlers. Also it has a decorator that takes the handler's returns and makes the http response
* Each handler has its own file with the implementation
* PersonStorage is separated from the handlers. It means that I used dependency injection with interfaces and instancing objects in order to make the code easier to change.
* It has inputs validations off request using go validator lib
* All is executed in `main.go` file

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,3 @@
+dev:
+	@echo "[dev] Running":
+	@nodemon --exec go run cmd/api/main.go --signal SIGTERM
```
Comment (cagodoy) [inline Makefile:8]: maybe you can find a other solution for hot reloading implemented in Golang, because `nodemon` script use `nodejs`.

By the other side, if I clone the project now and then I will run `make dev` command the script will be failed because I have not installed `nodemon` on my computer (and it's not specified in a `package.json` file or in install 
 make command.

File: pkg/api/create_person_handler.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package api
+
+import (
+	"encoding/json"
+
+	"github.com/darchlabs/api-example/pkg/person"
+	"github.com/go-playground/validator/v10"
+)
+
+type CreatePersonHandlerInputs struct {
```
Comment (cagodoy) [inline pkg/api/create_person_handler.go]: Maybe you must use the original `Person` type defined in the `Person` package because you can apply multiple metadata and you can centralize the person definition.

See the below photo:
 
Screen Shot 2022-10-17 at 15 31 43

File: pkg/api/create_person_handler.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+func (cp CreatePersonHandler) Invoke(ctx *handlerCtx) *handlerRes {
+	// Validate json schema data
+	validate := validator.New()
+	validate.Struct(ctx.ps)
+
+	// prepare struct for body
+	body := &struct {
+		Person *person.Person `json:"person"`
```
Comment (cagodoy) [inline pkg/api/create_person_handler.go:32]: Following your implementation y think that you use the `CreatePersonHandlerInputs` type definition

File: pkg/api/router.go [hunk 4]
```diff
@@ -1,21 +1,100 @@
...
+
+	"github.com/darchlabs/api-example/pkg/person"
 	"github.com/julienschmidt/httprouter"
 )
 
+// Define response structure
 type response struct {
-	Data interface{} `json:"data,omitempty"`
-	Meta interface{} `json:"meta,omitempty"`
+	Data  interface{} `json:"data,omitempty"`
```
Comment (cagodoy) [inline pkg/api/router.go:13]: remove this spaces

File: pkg/api/create_person_handler.go (hunk 2 above)
Comment (cagodoy) [inline pkg/api/create_person_handler.go]: I think that you do not use this type of implementation on anywhere of PR

File: pkg/api/create_person_handler.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+func NewCreatePersonHandler(ps PersonStorage) *CreatePersonHandler {
+	return &CreatePersonHandler{
+		storage: ps,
+	}
+}
+
+func (cp CreatePersonHandler) Invoke(ctx *handlerCtx) *handlerRes {
+	// Validate json schema data
+	validate := validator.New()
+	validate.Struct(ctx.ps)
```
Comment (cagodoy) [inline pkg/api/create_person_handler.go:25]: Are you sure that the code is using this validation? 
Did you try each edge case using postman? 

In this case, I think that you must implement a Test because you can try all different input cases.
The same for all handlers

File: pkg/api/update_person_handler.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	id := ctx.r.FormValue("id")
+
+	// Validate body
+	validate := validator.New()
+	validate.Struct(ctx.ps)
+
+	// Define structure for parse body
+	body := &struct {
+		Person *person.Person `json:"person"`
+	}{} // TODO(ca): What does this second '{}' means?
```
Comment (cagodoy) [inline pkg/api/update_person_handler.go]: The `TODO` should have the initials of the writer. In this case `TODO(nb): ...`

Answering your question is because it's an inline definition, you are defining the struct and at the same time, you are initializing it. In short words, the second `{}` is an initialization with empty values for person=`nil`.

File: pkg/storage/db.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+type DB struct {
+	DB *leveldb.DB
+}
+
+func New(filepath string) (*DB, error) {
+	db, err := leveldb.OpenFile(fmt.Sprintf("./%s", filepath), nil)
```
Comment (cagodoy) [inline pkg/storage/db.go:18]: missing validation for empty value or invalid value for `filepath` param

File: pkg/storage/person/storage.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (s *ps) Create(p *person.Person) (*person.Person, error) {
+	// generate id for database
+
```
Comment (cagodoy) [inline pkg/storage/person/storage.go:84]: remove this linebreak

File: pkg/storage/person/storage.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	return p, nil
+}
+
+func (s *ps) DeletePersonById(id string) (*person.Person, error) {
+	data, err := s.DB.DB.Get([]byte(id), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var pperson *person.Person
```
Comment (cagodoy) [inline pkg/storage/person/storage.go]: use `person` or `p` it's more clear.

File: pkg/storage/person/storage.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+	return p, nil
+}
+
+func (s *ps) GetPersonById(id string) (*person.Person, error) {
+	data, err := s.DB.DB.Get([]byte(id), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var pperson *person.Person
```
Comment (cagodoy) [inline pkg/storage/person/storage.go]: use `person` or `p` it's more clear.

File: pkg/storage/person/storage.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+	err = s.DB.DB.Put([]byte(id), b, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+func (s *ps) DeletePersonById(id string) (*person.Person, error) {
+	data, err := s.DB.DB.Get([]byte(id), nil)
```
Comment (cagodoy) [inline pkg/storage/person/storage.go:171]: missing validation for `id` param

File: pkg/storage/person/storage.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+	err = json.Unmarshal(data, &pperson)
+	if err != nil {
+		return nil, err
+	}
+
+	return pperson, err
+}
+
+func (s *ps) UpdatePersonById(id string, p *person.Person) (*person.Person, error) {
```
Comment (cagodoy) [inline pkg/storage/person/storage.go:131]: mising validations for params

File: pkg/storage/person/storage.go [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+
+	err := iter.Error()
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (s *ps) Create(p *person.Person) (*person.Person, error) {
```
Comment (cagodoy) [inline pkg/storage/person/storage.go:56]: missing validations for params

File: pkg/api/router.go [hunk 14]
```diff
@@ -1,21 +1,100 @@
...
+		handlerRes := fn(ctx)
+		payload, statusCode, err := handlerRes.Payload, handlerRes.HttpStatus, handlerRes.err
+		if err != nil {
+			// Response an error if sthing fails
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+
+		// Prepare response
+		res := response{
+			Meta: map[string]interface{}{"status_code": statusCode},
```
Comment (cagodoy) [inline pkg/api/router.go]: its neccesary define its the `json` response should be use `cammelCase` or `snakeCase`

**Review by cagodoy — CHANGES_REQUESTED** [review]

File: pkg/api/router.go (hunk 4 above)
Comment (nicobarra) [reply pkg/api/router.go:13]: Which spaces?

```// Define response structure
type response struct {
	Data  interface{} `json:"data,omitempty"`
	Meta  interface{} `json:"meta,omitempty"`
	Error interface{} `json:"error,omitempty"`
}
```

Here it is how I have those lines of codes right now.

File: Makefile (hunk 1 above)
Comment (nicobarra) [reply Makefile:8]: Ok

File: pkg/api/create_person_handler.go (hunk 2 above)
Comment (nicobarra) [reply pkg/api/create_person_handler.go]: Yeah, is better in that way. Done 👍

File: pkg/api/create_person_handler.go (hunk 3 above)
Comment (nicobarra) [reply pkg/api/create_person_handler.go:32]: But they are different, `CreatePersonHandlerInputs` contains the methods of `PersonStorage` struct, and `Person` is an struct with the fields required by the person object.
Maybe you say is better if I would pass `Person.person` instead of creating an initializing an struct with it?

File: pkg/api/create_person_handler.go (hunk 2 above)
Comment (nicobarra) [reply pkg/api/create_person_handler.go]: Got it, already changed

File: pkg/api/update_person_handler.go (hunk 6 above)
Comment (nicobarra) [reply pkg/api/update_person_handler.go]: Ok, good to know 👍

File: pkg/api/create_person_handler.go (hunk 3 above)
Comment (cagodoy) [reply pkg/api/create_person_handler.go:32]: I meant that you define the type but I don't see that you use it anywhere else, in fact I did the search in the entire PR and you don't use it anywhere

**Comment by nicobarra** [comment]
Last changes:
* Fixed requested improvements from @cagodoy (they are marked with a 👍)
* Created internal validation func's in personstorage for validating params
* Added assertion in personstorage Create func for failing if the person to create already exists
